a10_octavia/controller/worker/controller_worker.py

```diff
@@ -661,27 +662,105 @@ def delete_member(self, member_id):
+        set_u_ids = set(updated_member_ids)
+
+        current_member_ids = set_o_ids.union(set_u_ids)
+
+        current_members = [self._member_repo.get(db_apis.get_session(), id=mid)
+                           for mid in current_member_ids]
+
+        for mem in current_members:
+            # Rollback status to prevent pending state lock
+            self._member_repo.update(db_apis.get_session(), mem.id,
+                                     provisioning_status=constants.ACTIVE)
+            LOG.info("Member with id {} and ip {} and port {} slated for "
+                     "batch update have been set to ACTIVE state.".format(
+                         mem.id, mem.ip_address, mem.protocol_port))
+
+        new_members = [self._member_repo.get(db_apis.get_session(), id=mid)
+                       for mid in new_member_ids]
+        for mem in new_members:
+            self._member_repo.delete(db_apis.get_session(), mem.id)
+            current_member_ids.add(mem.id)
+            LOG.info("Member with id {} and ip {} and port {} "
+                     "slated for creation under batch update "
+                     "has been deleted.".format(mem.id, mem.ip_address, mem.protocol_port))
+
+        modified_members = current_members + new_members
+        for mem in modified_members:
+            if mem.pool_id is not None:
+                self._pool_repo.update(db_apis.get_session(), mem.pool_id,
+                                       provisioning_status=constants.ACTIVE)
+            if mem.pool.load_balancer_id is not None:
+                self._lb_repo.update(db_apis.get_session(),
+                                     mem.pool.load_balancer_id,
+                                     provisioning_status=constants.ACTIVE)
+
     @tenacity.retry(
         retry=tenacity.retry_if_exception_type(db_exceptions.NoResultFound),
         wait=tenacity.wait_incrementing(
             RETRY_INITIAL_DELAY, RETRY_BACKOFF, RETRY_MAX),
         stop=tenacity.stop_after_attempt(RETRY_ATTEMPTS))
     def batch_update_members(self, old_member_ids, new_member_ids,
-                             updated_members):
+                             updated_members_req):
 
+        updated_member_ids = [m.get('id') for m in updated_members_req]
+        updated_member_models = [self._member_repo.get(db_apis.get_session(), id=mid)
+                                 for mid in updated_member_ids]
+        old_members = [self._member_repo.get(db_apis.get_session(), id=mid)
+                       for mid in old_member_ids]
         new_members = [self._member_repo.get(db_apis.get_session(), id=mid)
                        for mid in new_member_ids]
+
+        modified_members = old_members + updated_member_models + new_members
+        member_collision_map = {}
+        for mem in modified_members:
+            mem_id = mem[0] if type(mem) == tuple else mem.id
+            if member_collision_map.get(mem_id):
+                member_collision_map[mem_id][1] += 1
```

Comment:
We definitely can. It is a trade off. If we move everything to within this for loop, we cut down on the linear portion of our runtime and space complexities. We may also cut down on DB transactions. However, it will impact how our log messages are printed to screen.

Current message order example:
````
<Log information on duplicate new>
<Log information on duplicate updated>
<Log rollback of duplicate update>
<Log rollback of duplicate new>
<Log warning about duplication occurring>
```

Altered message order example:
```
<Log information of duplicate new>
<Log rollback of duplicate new>
<Log information of duplicate update>
<Log roll back of duplicate update>
<Log warning about duplication occurring>
```

I figured that the first would be preferred from the operator standpoint since all of the information about the duplicates is in the same spot instead of being interleaved between rollback printouts.

Reply:
We can just rollback and raise though. We do have to provide information on the members otherwise hunting down the duplicates will be a tedious process for the operator who is working at scale of a hundred members. Even finding a duplicate port + IP combination within tens of members could be problematic.